src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs

```diff
@@ -1182,6 +1190,11 @@ private Task InternalExecuteNonQuery(TaskCompletionSource<object> completion, bo
             return task;
         }
 
+        private static Task InternalExecuteNonQuerySetupCloseContinuation(Task task, SqlDataReader reader)
+        {
+            return AsyncHelper.CreateContinuationTask(task, () => reader.Close());
```

Comment:
Why not change these to support passing in object state and having that passed into the delegate?  That'll avoid the need for closures entirely, and we can avoid creating all these little helper methods.  That said, you can avoid creating all these little helpers by just ensuring that the data being closed over is all in a local scope, as the closure will be created at the smallest scope necessary to encapsulate everything.

Reply:
Avoiding the closures sounds ideal, how would I go about changing it to do that?

Reply:
@stephentoub if I change this to use a state object which method of passing the state would be best? I see three obvious options, a specific type (extra type, clear on all param names), valuetuple (no extra type in code, boxes), tuple (no param names, no extra type in code)?

Reply:
I generally use a tuple or a value tuple, depending on whether it's going to be passed through as an object or as a strongly-typed value.  If the former, I use a tuple, if the latter, I use a value tuple.  That's assuming there's more than one value to pass through: if there's just the one, I just pass that single one as the state.